GPDB codegen utils

hardikar opened the pull request:
Utility module to support integrating LLVM code-generation in GPDB.
This augments the core functionality provided by the LLVM C++ libraries for program construction, optimization, and execution with additional features to help runtime-generated code integrate with statically compiled code of GPDB. There is additional information about the specific features in the [Codegen README](https://github.com/hardikar/gpdb/tree/uniform-cmake/src/backend/codegen).

In summary, you can build GPDB with codegen by:
``` ./configure --enable-codegen && make```
This depends on the presence of LLVM and Clang libraries and cmake as a build dependency . (Look at the README for more details). Note that none of these dependencies are applicable when  ```--enable-codegen``` is not specified.

All the methods in the utility modules are tested with unit tests written using gtest, giving good test coverage. You can run these tests (after configuring) by ```make -C src/backend/codegen unittest-check```.

For more information about the codegen project, please refer to the [OSS Codegen GPDB Proposal](https://docs.google.com/document/d/1MPyfvRWwnFREgriTwS2hkk_WHbvrTOW3gAtva_2epJ0/edit#heading=h.zadz5ppw5gmc).

Authors
@craig-chasseur, @armenatzoglou, @hardikar

hlinnaka commented:
* What does this actually do? AFAICS this is just infrastructure for 
something, and doesn't actually add any new functionality. It can be 
useful to review this as it is, but it's difficult to see whether the 
abstractions are sensible without seeing what this is going to be used 
for. We can discuss and review this now, but we're going to need some 
example code or at least an explanation of how this is going to be used, 
to do a good job at it. But please make sure to *not* push this, because 
as it is, it just adds a lot of unused code. (And the commit history 
will obviously need to be squashed to be readable, but that can wait 
until this is otherwise in a committable state)

* The rest of the codebase uses autoconf, not cmake. Please rewrite the 
makefiles using the autoconf+make, to avoid adding the new dependency to 
cmake, and to keep things simple.

* How is this going to be used? What's the API that this presents to the 
rest of the backend code?

* A lot of the files have an Apache license header. Does this code come 
from some 3rd party source, under the Apache license, or is this all 
pivotal-copyrighted?

* Please try to keep the C++ portion of the module to the minimum, and 
keep it well separated from the rest of the codebase. I'm a bit worried 
about having .h files in src/include/codegen that are actually C++ 
headers, for example. They're only going to be usable within the codegen 
module, so perhaps it would be better to have a separate include 
directory within src/backend/codegen for those, and only expose the API 
that's going to be called from the rest of the backend code as .h files 
in src/include. (And should they use the .hpp extension, as they're C++ 
headers? I'm no expert in C++...)

- Heikki

vraghavan78 commented:
* @hlinnaka: Cmake vs. Autoconf this may be a matter of opinion. IMHO cmake is both easier, and manageable means of making. In early 2015, the make system inside Orca was similar to GPDB/Postgres. It was awful and unnecessarily complex that can be achieved in a few lines via CMake! CMake is so simple, clean and comes with the benefit of CTest. 

* Ideally the entire code base will be in CMake, but that will cause grief to folks who are merging, so no need it for it now.

* Codegen similar to Orca header files. So what is the difference here? Is your aversion just C++? With Orca we have shown how a C++ module can work cleanly in a production system.

* In regards to the unnecessary dependency, if the open source user wants Orca, they need CMake anyways.

hlinnaka commented:
On 19/03/16 10:57, Venkatesh Raghavan wrote:
> * @hlinnaka: Cmake vs. Autoconf this may be a matter of opinion. IMHO
> cmake is both easier, and manageable means of making. In early 2015,
> the make system inside Orca was similar to GPDB/Postgres. It was
> awful and unnecessarily complex that can be achieved in a few lines
> via CMake! CMake is so simple, clean and comes with the benefit of
> CTest.

ORCA is a completely separate project, while codegen is just a 
subdirectory inside the Greenplum backend directory. It's very different 
situation.

All you need for codegen is a Makefile within src/backend/codegen, like 
the Makefiles for other subdirectories, and checks in configure.in for 
the LLVM libs and headers. Perhaps you'll need a few more flags there 
for C++ options, but that's it.

> * Ideally the entire code base will be in CMake, but that will cause
> grief to folks who are merging, so no need it for it now.

For better or worse, we're stuck with autoconf. The question is: is it 
simpler to use autoconf for most parts of the system, and cmake for 
others, and ensure that both of them consistently produce the same 
result wrt. the libraries and compiler options that they choose, or just 
use autoconf for everything.

> * Codegen similar to Orca header files. So what is the difference
> here? Is your aversion just C++? With Orca we have shown how a C++
> module can work cleanly in a production system.

I have an aversion for using C++ *in PostgreSQL code*, because C++ does 
not interact well with the Postgres idioms like exception handling and 
memory management. Yes, it's possible, as Orca has shown, and IIRC 
PostGIS does similar things, but it's a helluva lot more complicated to 
integrate C++ code into the backend robustly than plain C code.

Yeah, to use LLVM in the backend, you pretty much have to use C++. So 
it's important to make that as robust as possible, and that means 
carefully controlling the interface between the C++ parts and the rest 
of the backend. That's what I was getting at with the headers.

This is indeed a very similar situation as with Orca. There are indeed a 
bunch of C++ header files in src/include/gpopt, and I think it has the 
same problem that it's not really clear which functions form the 
external API of gpopt, and which ones are private to it. What is the 
entry point into Orca? Which .h file or files in src/include/gpopt are 
supposed to be #included from the rest of the backend, and which are C++ 
headers internal to Orca/gpopt? By doing "ls -l src/include/gpopt", it's 
not obvious.

Granted, it's not a big problem, and you could also help that by better 
comments or naming the header files differently, for example. Separating 
the C++ headers that are internal to codegen from C headers that form 
the interface to the rest of the backend is another thing that helps 
with that.

> * In regards to the unnecessary dependency, if the open source user
> wants Orca, they need CMake anyways.

Ok, but that has nothing to do with codegen.

- Heikki

danielgustafsson commented:
Repositories outside GPDB can use other build tools but for GPDB.git, and all the code in it, a single top-level `./configure` and `make` should be all the commands and tools needed for building all the parts of GPDB you want/need. I'm against introducing a dependency on CMake regardless of how good CMake may be.

ascherbaum-pivotal commented:
> Repositories outside GPDB can use other build tools but for GPDB.git, and all the code in it, a single top-level ./configure and make should be all the commands and tools needed for building all the parts of GPDB you want/need.

To add to this note:
I see that configure and configure.in define a $CMAKE_CMD, but there is no test if "cmake" is even available. As a result, the build will fail in a subdirectory where cmake is used.

Not good.

gcaragea commented:
On Sat, Mar 19, 2016 at 1:31 AM, Heikki Linnakangas <
notifications@github.com> wrote:

> * What does this actually do? AFAICS this is just infrastructure for
> something, and doesn't actually add any new functionality. It can be
> useful to review this as it is, but it's difficult to see whether the
> abstractions are sensible without seeing what this is going to be used
> for. We can discuss and review this now, but we're going to need some
> example code or at least an explanation of how this is going to be used,
> to do a good job at it. But please make sure to *not* push this, because
> as it is, it just adds a lot of unused code. (And the commit history
> will obviously need to be squashed to be readable, but that can wait
> until this is otherwise in a committable state)
>
>
We are working on a first full-working prototype, but that will take us a
few more days, and it requires changes to GPDB executor code. We have
simple examples that use the API, but they were not beneficial enough (in
terms of performance) to be pushed to GPDB just yet. Please refer to the OSS
Codegen GPDB Proposal
<https://docs.google.com/document/d/1MPyfvRWwnFREgriTwS2hkk_WHbvrTOW3gAtva_2epJ0/edit#heading=h.zadz5ppw5gmc>
for some use cases.

We opted to operate in a way where we want to get early feedback on the
code, and we'd like to push this code in as a framework, rather than wait
another couple of weeks and let the code sit for a long time in a separate
branch. Please note that the framework proposed today is fully tested
through unit tests. There is a configure flag that completely disables this
feature (a feature flag), so this shouldn't affect any other developers or
tests at this point.

We'll proceed in an iterative way, where we'll start pushing use cases
where we generate code for various hot-spots in GPDB, one at a time.

(We'll squash the commits before pushing. The only reason we kept the
entire history in this PR is because you suggested to do that, to help in
the review).

Thanks for reviewing.
George

gcaragea commented:
On Sat, Mar 19, 2016 at 1:31 AM, Heikki Linnakangas <
notifications@github.com> wrote:

>
> * How is this going to be used? What's the API that this presents to the
> rest of the backend code?
>
>
Please refer to the OSS Codegen GPDB Proposal
<https://docs.google.com/document/d/1MPyfvRWwnFREgriTwS2hkk_WHbvrTOW3gAtva_2epJ0/edit#heading=h.zadz5ppw5gmc>
for the API and uses cases. We discuss it there.


> * A lot of the files have an Apache license header. Does this code come
> from some 3rd party source, under the Apache license, or is this all
> pivotal-copyrighted?
>

We'll address the license, thanks for pointing that out. There isn't any
3rd party source code in the utils part, it was written from scratch.


>
> * Please try to keep the C++ portion of the module to the minimum, and
> keep it well separated from the rest of the codebase. I'm a bit worried
> about having .h files in src/include/codegen that are actually C++
> headers, for example. They're only going to be usable within the codegen
> module, so perhaps it would be better to have a separate include
> directory within src/backend/codegen for those, and only expose the API
> that's going to be called from the rest of the backend code as .h files
> in src/include. (And should they use the .hpp extension, as they're C++
> headers? I'm no expert in C++...)
>
>
Sounds good, I like this suggestion, it should be possible to do. We'll
address this, thanks.

George

hlinnaka commented:
On 21/03/16 20:28, George Caragea wrote:
> We are working on a first full-working prototype, but that will take us a
> few more days, and it requires changes to GPDB executor code. We have
> simple examples that use the API, but they were not beneficial enough (in
> terms of performance) to be pushed to GPDB just yet. Please refer to the OSS
> Codegen GPDB Proposal
> <https://docs.google.com/document/d/1MPyfvRWwnFREgriTwS2hkk_WHbvrTOW3gAtva_2epJ0/edit#heading=h.zadz5ppw5gmc>
> for some use cases.
>
> We opted to operate in a way where we want to get early feedback on the
> code, and we'd like to push this code in as a framework, rather than wait
> another couple of weeks and let the code sit for a long time in a separate
> branch. Please note that the framework proposed today is fully tested
> through unit tests. There is a configure flag that completely disables this
> feature (a feature flag), so this shouldn't affect any other developers or
> tests at this point.

Yeah, I'm glad we got to start the discussion on the build system early. 
That said, this is new code, so it's not going to bitrot. There's no 
hurry in pushing it, so please don't.

Without seeing the code that's going to use this, it's impossible to 
tell how much of this framework is actually needed, which parts of it 
are important, etc.

> We'll proceed in an iterative way, where we'll start pushing use cases
> where we generate code for various hot-spots in GPDB, one at a time.

Let's take an incremental approach and get the simplest possible code 
generation actually working with this, and include only those parts of 
the framework that are needed to support that. Let's see what the 
generation of slot_deform_tuple() would look like with this. That will 
allow us to look at the API, the frameworks needed, the testing needs 
and so on. Until then, we're just speculating what we'll need.

- Heikki

vraghavan78 commented:
>>> There's no hurry in pushing it, so please don't.

We do not want to have a long living feature branches. 

Till we have the initial prototype from GPDB's perspective this is code is not exercised. So I do not see the issue with pushing it into master. Am I missing something?

hlinnaka commented:
On 21/03/16 22:12, Venkatesh Raghavan wrote:
>>>> There's no hurry in pushing it, so please don't.
>
> We do not want to have a long living feature branches.

There is no feature in this yet. It's just a bunch of functions, which 
we might or might not need.

> Till we have the initial prototype from GPDB's perspective this is
> code is not exercised. So I do not see the issue with pushing it into
> master. Am I missing something?

Till we have an initial prototype that uses this code, it is dead code 
and there is no point in having it in the repo.

- Heikki

foyzur commented:
I think we all want the best for GPDB and just have different perspective on what is the best way to get there :-) Long living branch is an absolute no no from our past experience. An iterative development depends on writing building blocks and pushing those into the master branch. The code is NOT a dead code because it has plenty of tests that exercises it. Not just that, by just looking at the tests we can easily figure out their use cases and future direction. The GPDB integration is just consumption of the framework from another client.

vraghavan78 commented:
>> There is no feature in this yet. It's just a bunch of functions, which we might or might not need.

We are building the ground work for it. The team has to balance customer escalations and looking ahead w.r.t. new features. From that point of view, I disagree with the point of view of not pushing it into master.

danielgustafsson commented:
What is to be gained from having this is in the tree before it is used? I don't see the problem of working on a feature like this in a fork until working code is ready and a neat and squashed set of commits can be reviewed and committed.

edespino commented:
+1 (be sure to squash prior to committing)

hardikar commented:
>>
> * Please try to keep the C++ portion of the module to the minimum, and
> keep it well separated from the rest of the codebase. I'm a bit worried
> about having .h files in src/include/codegen that are actually C++
> headers, for example. They're only going to be usable within the codegen
> module, so perhaps it would be better to have a separate include
> directory within src/backend/codegen for those, and only expose the API
> that's going to be called from the rest of the backend code as .h files
> in src/include. (And should they use the .hpp extension, as they're C++
> headers? I'm no expert in C++...)
>
>
>Sounds good, I like this suggestion, it should be possible to do. We'll
>address this, thanks.

Agreed. We have moved the codegen utils related header files into the src/backend/codegen/include directory.

>> * A lot of the files have an Apache license header. Does this code come
>> from some 3rd party source, under the Apache license, or is this all
>> pivotal-copyrighted?
>>
>
>We'll address the license, thanks for pointing that out. There isn't any
>3rd party source code in the utils part, it was written from scratch.

Thanks for pointing this out. We have removed the Apache license headers, and replaced it with pivotal copyright headers.

hardikar commented:
>> Repositories outside GPDB can use other build tools but for GPDB.git, and all the code in it, a single top-level ./configure and make should be all the commands and tools needed for building all the parts of GPDB you want/need.

>To add to this note:
>I see that configure and configure.in define a $CMAKE_CMD, but there is no test if "cmake" is even available. As a result, the build will fail in a subdirectory where cmake is used.

>Not good.

To clarify, the build doesn't fail, but the configure does fail, giving you a message saying "cmake: command not found" - and this happens only when you --enable-codegen. But we agree, we should be checking that cmake is present in the path. We added a check in configure for the existence of cmake.

danielgustafsson commented on src/backend/codegen/README.md:
````diff
@@ -0,0 +1,255 @@
+default locations, you will need to point cmake at it when building codegen
+like so:
+```
+./configure --enable-codegen CMAKE_PREFIX_PATH=/usr/local/opt/llvm37/lib/llvm-3.7
````
To match the style of the top level autoconf this should be a `--with-llvm=DIR` flag rather than an environment variable.

hardikar replied:
That will make it more consistent, however we must find a way to give the prefix path of llvm and clang. CMAKE_PREIFX_PATH takes a list of directories separated by colons. Would ```--with-codegen-prefix=DIRECTORIES``` work better?

hlinnaka replied:
On 23/03/16 21:07, Shreedhar Hardikar wrote:
> That will make it more consistent, however we must find a way to give
> the prefix path of llvm and clang. CMAKE_PREIFX_PATH takes a list of
> directories separated by colons. Would
> ```--with-codegen-prefix=DIRECTORIES``` work better?

Or, we could just keep things simple and use autoconf directly...

- Heikki

danielgustafsson commented on src/backend/codegen/README.md:
````diff
@@ -0,0 +1,255 @@
+brew install --with-clang llvm37
+```
+
+Unfortunately, the cmake configuration module for LLVM that is isntalled by
````
s/isntalled/installed/

danielgustafsson commented:
I'm still not in favour of adding a build dependency on cmake (the question of cmake being good/bad is not the point) since afaict it will defer discovery of library dependencies etc from the configure step into the build step. What was the issue encountered with using the top-level autoconf file? Looking at the cmake files they seem pretty simple to port?

danielgustafsson commented:
The amount of commits in this PR is overwhelming to review, can you please squash the commits into what you propose to merge into the tree to make it easier for reviewing?

hardikar commented:
>I'm still not in favour of adding a build dependency on cmake (the question of cmake being good/bad is not the point) since afaict it will defer discovery of library dependencies etc from the configure step into the build step.
Cmake itself is a two step system (much like configure+make), and we've linked up the the first step of cmake with configure step of GPDB. In other words, the configure step will fail (and should) if there are any libraries missing. If you see a different behavior, we should definitely get that fixed. 

>The amount of commits in this PR is overwhelming to review, can you please squash the commits into what you propose to merge into the tree to make it easier for reviewing?
Yes, working on that presently.

armenatzoglou commented:
Pushed to master (88e9baba520d1bdb2f3686ba7257660f45d44086). Thanks a lot for reviewing.
